### PR TITLE
AuthN: Add gRPC authenticator

### PR DESCRIPTION
**Pull request by gamab**
This PR introduces a new gRPC authenticator that verifies incoming requests based on access and ID tokens.

**Key changes:**

* Adds `GrpcAuthenticator` that performs authentication based on access and ID tokens present in the request metadata.
* Verifies the tokens using existing `Verifier` implementations for both access and ID tokens.
* Extracts the stack ID from the request metadata and ensures both tokens belong to the expected namespace.
* Validates the token subject types, allowing only specific types for access and ID tokens.
* Returns a context enriched with caller information upon successful authentication.

**TODOs:**

* Implement constructor for `GrpcAuthenticator`.
* Make ID token verification optional.
* Allow configuration of metadata keys for tokens and stack ID.
* Implement configurable stack ID extraction from various sources (metadata, path, ID token).

**Next Steps:**

* Integrate the `GrpcAuthenticator` into the gRPC server pipeline.
* Configure the authenticator with appropriate verifiers and customization options.
* Update documentation and examples to showcase the new authentication mechanism.

This change enhances the security of our gRPC services by providing a robust and flexible authentication mechanism.

### CODE REVIEW

File: authn/grpc_authenticator.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+	return values[0], true
+}
+
+// ------
+// Subject
+// FIXME: This is a duplicate of Grafana's identity namespaces and namespacedID. It should be moved to a shared package.
+// ------
+const (
+	namespaceUser           subjectNamespace = "user"
```
Comment (gamab) [inline authn/grpc_authenticator.go:153]: To me, using the `namespace` keyword is super confusing as it conflicts with the token `namespace`. I'd prefer `subjectType` or `subjectKind`

File: authn/grpc_authenticator.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	if len(values[0]) == 0 {
+		return "", false
+	}
+
+	return values[0], true
+}
+
+// ------
+// Subject
+// FIXME: This is a duplicate of Grafana's identity namespaces and namespacedID. It should be moved to a shared package.
```
Comment (gamab) [inline authn/grpc_authenticator.go:149]: My apologies, I went quick and dirty there. I know Kalle wanted to migrate the authn identity code here **properly**.

File: authn/grpc_authenticator.go (hunk 1 above)
Comment (gamab) [reply authn/grpc_authenticator.go]: Change done

File: authn/grpc_authenticator.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+// ------
+const (
+	typeUser           subjectType = "user"
```
Comment (Jguer) [inline authn/grpc_authenticator.go:152]: nice

**Review by Jguer — APPROVED** [review]

**Review by colin-stuart — APPROVED** [review]

**Pull request merged**